File: src/main/java/de/pflugradts/passbird/domain/service/eventhandling/DomainEventHandler.java
package de.pflugradts.passbird.domain.service.eventhandling;

import com.google.common.eventbus.Subscribe;
import com.google.inject.Inject;
import com.google.inject.Singleton;
import de.pflugradts.passbird.domain.model.PasswordEntryDiscarded;
import de.pflugradts.passbird.domain.service.password.storage.PasswordEntryRepository;

@Singleton
public class DomainEventHandler implements EventHandler {

    @Inject
    private PasswordEntryRepository passwordEntryRepository;

    @Subscribe
    private void handlePasswordEntryDiscarded(final PasswordEntryDiscarded passwordEntryDiscarded) {
        passwordEntryRepository.delete(passwordEntryDiscarded.getPasswordEntry());
    }

}

File: src/main/kotlin/de/pflugradts/passbird/domain/service/eventhandling/DomainEventHandler.kt
package de.pflugradts.passbird.domain.service.eventhandling

import de.pflugradts.passbird.domain.model.ddd.DomainEvent
import de.pflugradts.passbird.domain.model.event.EggDiscarded
import de.pflugradts.passbird.domain.service.password.tree.EggRepository

class DomainEventHandler(private val eggRepositoryProvider: () -> EggRepository) : EventHandler {
    override val eventTypes: Set<Class<out DomainEvent>> = setOf(EggDiscarded::class.java)

    private val eggRepository: EggRepository get() = eggRepositoryProvider()

    override fun handle(domainEvent: DomainEvent) {
        when (domainEvent) {
            is EggDiscarded -> handleEggDiscarded(domainEvent)
        }
    }

    private fun handleEggDiscarded(eggDiscarded: EggDiscarded) {
        eggRepository.delete(eggDiscarded.egg)
    }
}
